src/rdb.c

```diff
@@ -2291,111 +2293,78 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, redisDb* db, int rdbflags,
                 {
                     /* convert to hash */
-                    hashTypeConvert(o, OBJ_ENCODING_HT, hexpires);
+                    hashTypeConvert(o, OBJ_ENCODING_HT, NULL);
 
                     if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
                         if (dictTryExpand(o->ptr, len) != DICT_OK) {
                             rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
                             decrRefCount(o);
                             if (dupSearchDict != NULL) dictRelease(dupSearchDict);
                             sdsfree(value);
-                            sdsfree(field);
+                            hfieldFree(field);
                             return NULL;
                         }
                     }
 
                     /* don't add the values to the new hash: the next if will catch and the values will be added there */
                 } else {
-                    listpackExAddNew(o, field, value, expireAt);
-                    sdsfree(field);
+                    listpackExAddNew(o, field, hfieldlen(field),
+                                     value, sdslen(value), expireAt);
+                    hfieldFree(field);
                     sdsfree(value);
                 }
             }
 
             if (o->encoding == OBJ_ENCODING_HT) {
-                /* WA for check-rdb mode, when there's no DB so can't attach expired items to ebuckets,
-                 * or when no expiry was not set for this field */
-                if ((db == NULL) || (expireAt == 0)) {
-                    hashTypeSet(db, o, field, value, 0);
-                } else {
-                    if (hashTypeSetExRdb(db, o, field, value, expireAt) != C_OK) {
-                        serverLog(LL_WARNING, "failed adding hash field %s to key %s", field, key);
-                        decrRefCount(o);
-                        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
-                        sdsfree(value);
-                        sdsfree(field);
-                        return NULL;
-                    }
+                /* Add pair to hash table */
+                dict *d = o->ptr;
+                dictUseStoredKeyApi(d, 1);
+                int ret = dictAdd(d, field, value);
+                dictUseStoredKeyApi(d, 0);
+
+                /* Attach expiry to the hash field and register in hash private HFE DS */
+                if ((ret != DICT_ERR) && expireAt) {
+                    dictExpireMetadata *m = (dictExpireMetadata *) dictMetadata(d);
+                    ret = ebAdd(&m->hfe, &hashFieldExpireBucketsType, field, expireAt);
+                }
+
+                if (ret == DICT_ERR) {
+                    rdbReportCorruptRDB("Duplicate hash fields detected");
+                    sdsfree(value);
```

Comment:
don't you need to free dupSearchDict here? or somewhere else before reaching to this point?

Reply:
There is an implicit assumption that once the object is `OBJ_ENCODING_HT`, then `dupSearchDict` got released.

Reply:
Yeah but looks like it is not the case. It is hard to follow this function but on line 2334, it converts to HT but does not free dupSearchDict.